CA1717 false positives for "status"

#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers

#### Analyzer

CA1717

#### Repro steps

Add a non-flags enumeration with a suffix of "Status", such as `ViewStatus`

#### Expected behavior

No violation, since "status" is not generally accepted as a pluralization.

#### Actual behavior
```
warning CA1717: Only FlagsAttribute enums should have plural names
```

--- a/THIRD-PARTY-NOTICES.txt
+++ b/THIRD-PARTY-NOTICES.txt
@@ -1,0 +1,43 @@
+Roslyn-Analyzers uses third-party libraries or other resources that may be
+distributed under licenses different than the Roslyn-Analyzers software.
+
+In the event that we accidentally failed to list a required notice, please
+bring it to our attention. Post an issue or email us:
+
+           dotnet@microsoft.com
+
+The attached notices are provided for information only.
+
+
+License notice for MSBuild Locator
+-------------------------------------
+
+https://github.com/Microsoft/MSBuildLocator
+
+Copyright (c) 2018 .NET Foundation and Contributors
+
+This software is licensed subject to the MIT license, available at 
+https://opensource.org/licenses/MIT
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+License notice for Humanizer
+-------------------------------------
+
+https://github.com/Humanizr/Humanizer
+
+Copyright (c) 2018 .NET Foundation and Contributors
+
+This software is licensed subject to the MIT license, available at 
+https://opensource.org/licenses/MIT
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -32,6 +32,7 @@
     <SystemReflectionMetadataVersion>1.4.2</SystemReflectionMetadataVersion>
     <MicrosoftVisualBasicVersion>10.1.0</MicrosoftVisualBasicVersion>
     <MicrosoftBuildLocatorVersion>1.1.2</MicrosoftBuildLocatorVersion>
+    <HumanizerVersion>2.5.16</HumanizerVersion>
   </PropertyGroup>
 
   <PropertyGroup>

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -32,7 +32,7 @@
     <SystemReflectionMetadataVersion>1.4.2</SystemReflectionMetadataVersion>
     <MicrosoftVisualBasicVersion>10.1.0</MicrosoftVisualBasicVersion>
     <MicrosoftBuildLocatorVersion>1.1.2</MicrosoftBuildLocatorVersion>
-    <HumanizerVersion>2.5.16</HumanizerVersion>
+    <HumanizerVersion>2.2.0</HumanizerVersion>
   </PropertyGroup>
 
   <PropertyGroup>

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EnumsShouldHavePluralNames.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EnumsShouldHavePluralNames.cs
@@ -130,6 +130,12 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 return;
             }
 
+            if (!symbol.Name.IsASCII())
+            {
+                // Skip non-ASCII names.
+                return;
+            }
+
             bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
             if (hasFlagsAttribute)
             {

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EnumsShouldHavePluralNames.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EnumsShouldHavePluralNames.cs
@@ -1,9 +1,12 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Immutable;
+using System.Globalization;
 using System.Linq;
 using Analyzer.Utilities;
 using Analyzer.Utilities.Extensions;
+using Humanizer;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
 
@@ -90,6 +93,13 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            if (!CultureInfo.CurrentCulture.Name.Equals("en", StringComparison.Ordinal) &&
+				!CultureInfo.CurrentCulture.Parent.Name.Equals("en", StringComparison.Ordinal))
+            {
+                // FxCop compat: Skip for non-English cultures.
+                return;
+            }
+
             analysisContext.EnableConcurrentExecution();
             analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 
@@ -113,22 +123,32 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                 return;
             }
 
+            if (symbol.Name.EndsWith("i", StringComparison.OrdinalIgnoreCase) || symbol.Name.EndsWith("ae", StringComparison.OrdinalIgnoreCase))
+            {
+                // Skip words ending with 'i' and 'ae' to avoid flagging irregular plurals.
+                // Humanizer does not recognize these as plurals, such as 'formulae', 'trophi', etc.
+                return;
+            }
+
             bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
             if (hasFlagsAttribute)
             {
-                if (!symbol.Name.IsPlural()) // Checking Rule CA1714
+                if (!IsPlural(symbol.Name)) // Checking Rule CA1714
                 {
                     context.ReportDiagnostic(symbol.CreateDiagnostic(Rule_CA1714, symbol.OriginalDefinition.Locations.First(), symbol.Name));
                 }
             }
             else
             {
-                if (symbol.Name.IsPlural()) // Checking Rule CA1717
+                if (IsPlural(symbol.Name)) // Checking Rule CA1717
                 {
                     context.ReportDiagnostic(symbol.CreateDiagnostic(Rule_CA1717, symbol.OriginalDefinition.Locations.First(), symbol.Name));
                 }
             }
         }
+
+        private static bool IsPlural(string word)
+            => word.Equals(word.Pluralize(inputIsKnownToBeSingular: false), StringComparison.OrdinalIgnoreCase);
     }
 }
 

--- a/src/Microsoft.CodeQuality.Analyzers/Core/Microsoft.CodeQuality.Analyzers.csproj
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/Microsoft.CodeQuality.Analyzers.csproj
@@ -12,6 +12,7 @@
     <InternalsVisibleTo Include="Microsoft.CodeQuality.CSharp.Analyzers" />
     <InternalsVisibleTo Include="Microsoft.CodeQuality.VisualBasic.Analyzers" />
     <InternalsVisibleTo Include="Microsoft.CodeQuality.Analyzers.UnitTests" />
+    <PackageReference Include="Humanizer.Core" Version="$(HumanizerVersion)"/>
   </ItemGroup>
   <Import Project="..\..\Utilities\Analyzer.Utilities.projitems" Label="Shared" />
 </Project>

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumsShouldHavePluralNamesTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumsShouldHavePluralNamesTests.cs
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Testing;
 using Test.Utilities;
 using Xunit;
 
@@ -60,7 +61,7 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests
                                        
                                 };
                             }",
-                            GetCSharpResultAt(4, 44, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717.MessageFormat.ToString()));
+                            GetCSharpNoPluralResultAt(4, 44));
 
             VerifyBasic(@"
                         Public Class A
@@ -72,7 +73,7 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests
 	                        End Enum
                         End Class
                         ",
-                        GetBasicResultAt(3, 38, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717.MessageFormat.ToString()));
+                        GetBasicNoPluralResultAt(3, 38));
         }
 
         [Fact, WorkItem(1432, "https://github.com/dotnet/roslyn-analyzers/issues/1432")]
@@ -154,7 +155,7 @@ End Class
                                        
                                 };
                             }",
-                            GetCSharpResultAt(4, 44, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717.MessageFormat.ToString()));
+                            GetCSharpNoPluralResultAt(4, 44));
 
             VerifyBasic(@"
                         Public Class A
@@ -166,7 +167,7 @@ End Class
 	                        End Enum
                         End Class
                         ",
-                        GetBasicResultAt(3, 38, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1717.MessageFormat.ToString()));
+                        GetBasicNoPluralResultAt(3, 38));
         }
 
         [Fact]
@@ -184,7 +185,7 @@ End Class
                                        
                                 };
                             }",
-                            GetCSharpResultAt(5, 44, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1714.MessageFormat.ToString()));
+                            GetCSharpPluralResultAt(5, 44));
 
             VerifyBasic(@"
                        Public Class A
@@ -195,7 +196,7 @@ End Class
 		                    Tuesday = 2
 	                    End Enum
                         End Class",
-                            GetBasicResultAt(4, 34, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1714.MessageFormat.ToString()));
+                            GetBasicPluralResultAt(4, 34));
         }
 
         [Fact, WorkItem(1432, "https://github.com/dotnet/roslyn-analyzers/issues/1432")]
@@ -323,8 +324,8 @@ End Class
                         End Class");
         }
 
-        [Fact]
-        public void CA1714_CA1717_Test_EnumWithFlags_NameEndsWithS()
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
+        public void CA1714_CA1717_Test_EnumWithFlags_NonPluralNameEndsWithS()
         {
             VerifyCSharp(@" 
                             public class A 
@@ -337,7 +338,8 @@ End Class
                                     z = 2
                                        
                                 };
-                            }");
+                            }",
+                            GetCSharpPluralResultAt(5, 44));
 
             VerifyBasic(@"
                        Public Class A
@@ -347,12 +349,38 @@ End Class
 		                    y = 1
 		                    z = 2
 	                    End Enum
+                        End Class",
+                        GetBasicPluralResultAt(4, 34));
+        }
+
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
+        public void CA1714_CA1717_Test_EnumWithFlags_PluralNameEndsWithS()
+        {
+            VerifyCSharp(@" 
+                            public class A 
+                            { 
+                               [System.Flags] 
+                               public enum Axes 
+                               {
+                                    x = 0,
+                                    y = 1,
+                                    z = 2
+                                       
+                                };
+                            }");
+
+            VerifyBasic(@"
+                       Public Class A
+	                    <System.Flags> _
+	                    Public Enum Axes
+		                    x = 0
+		                    y = 1
+		                    z = 2
+	                    End Enum
                         End Class");
         }
 
-        [Fact]
-        // TODO: when the Isplural function handles better checks.
-        // this test should not fail.
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
         public void CA1714_CA1717_Test_EnumWithFlags_PluralName_NotEndingWithS()
         {
             VerifyCSharp(@" 
@@ -366,8 +394,7 @@ End Class
                                     M3 = 2
                                        
                                 };
-                            }",
-                            GetCSharpResultAt(5, 44, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1714.MessageFormat.ToString()));
+                            }");
 
             VerifyBasic(@"
                        Public Class A
@@ -377,11 +404,10 @@ End Class
                             M2 = 1
                             M3 = 2
                         End Enum
-                        End Class",
-                            GetBasicResultAt(4, 37, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, EnumsShouldHavePluralNamesAnalyzer.Rule_CA1714.MessageFormat.ToString()));
+                        End Class");
         }
 
-        [Fact]
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
         public void CA1714_CA1717_Test_EnumWithNoFlags_PluralWord_NotEndingWithS()
         {
             VerifyCSharp(@" 
@@ -394,7 +420,8 @@ End Class
                                     M3 = 2
                                        
                                 };
-                            }");
+                            }",
+                            GetCSharpNoPluralResultAt(4, 44));
 
             VerifyBasic(@"
                        Public Class A
@@ -403,12 +430,14 @@ End Class
                             M2 = 1
                             M3 = 2
                         End Enum
-                        End Class");
+                        End Class",
+                        GetBasicNoPluralResultAt(3, 37));
         }
 
-        [Fact]
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
         public void CA1714_CA1717_Test_EnumWithNoFlags_irregularPluralWord_EndingWith_ae()
         {
+            // Humanizer does not recognize 'formulae' as plural, but we skip words ending with 'ae'
             VerifyCSharp(@" 
                             public class A 
                             { 
@@ -418,7 +447,7 @@ End Class
                                     M1 = 0,
                                     M2 = 1,
                                     M3 = 2
-                                       
+
                                 };
                             }");
 
@@ -432,9 +461,11 @@ End Class
                         End Enum
                         End Class");
         }
-        [Fact]
+
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
         public void CA1714_CA1717_Test_EnumWithNoFlags_irregularPluralWord_EndingWith_i()
         {
+            // Humanizer does not recognize 'trophi' as plural, but we skip words ending with 'i'
             VerifyCSharp(@" 
                             public class A 
                             { 
@@ -457,6 +488,26 @@ End Class
                             M3 = 2
                         End Enum
                         End Class");
+        }
+
+        private static DiagnosticResult GetCSharpPluralResultAt(int line, int column)
+        {
+            return GetCSharpResultAt(line, column, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, MicrosoftApiDesignGuidelinesAnalyzersResources.FlagsEnumsShouldHavePluralNamesMessage);
+        }
+
+        private static DiagnosticResult GetBasicPluralResultAt(int line, int column)
+        {
+            return GetBasicResultAt(line, column, EnumsShouldHavePluralNamesAnalyzer.RuleId_Plural, MicrosoftApiDesignGuidelinesAnalyzersResources.FlagsEnumsShouldHavePluralNamesMessage);
+        }
+
+        private static DiagnosticResult GetCSharpNoPluralResultAt(int line, int column)
+        {
+            return GetCSharpResultAt(line, column, EnumsShouldHavePluralNamesAnalyzer.RuleId_NoPlural, MicrosoftApiDesignGuidelinesAnalyzersResources.OnlyFlagsEnumsShouldHavePluralNamesMessage);
+        }
+
+        private static DiagnosticResult GetBasicNoPluralResultAt(int line, int column)
+        {
+            return GetBasicResultAt(line, column, EnumsShouldHavePluralNamesAnalyzer.RuleId_NoPlural, MicrosoftApiDesignGuidelinesAnalyzersResources.OnlyFlagsEnumsShouldHavePluralNamesMessage);
         }
     }
 }

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumsShouldHavePluralNamesTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumsShouldHavePluralNamesTests.cs
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Testing;
@@ -483,6 +483,34 @@ End Class
                        Public Class A
                         < System.Flags > _
                         Public Enum trophi
+                            M1 = 0
+                            M2 = 1
+                            M3 = 2
+                        End Enum
+                        End Class");
+        }
+
+        [Fact, WorkItem(1323, "https://github.com/dotnet/roslyn-analyzers/issues/1323")]
+        public void CA1714_CA1717_Test_EnumWithNoFlags_NonAscii()
+        {
+            // We skip non-ASCII names.
+            VerifyCSharp(@" 
+                            public class A 
+                            { 
+                               [System.Flags] 
+                               public enum UnicodeNameΔ
+                               {
+                                    M1 = 0,
+                                    M2 = 1,
+                                    M3 = 2
+
+                                };
+                            }");
+
+            VerifyBasic(@"
+                       Public Class A
+                        < System.Flags > _
+                        Public Enum UnicodeNameΔ
                             M1 = 0
                             M2 = 1
                             M3 = 2

--- a/src/Test.Utilities/Test.Utilities.csproj
+++ b/src/Test.Utilities/Test.Utilities.csproj
@@ -17,6 +17,7 @@
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.XUnit" Version="$(MicrosoftCodeAnalysisTestingVersion)" />
     <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.CodeFix.Testing.XUnit" Version="$(MicrosoftCodeAnalysisTestingVersion)" />
     <PackageReference Include="xunit.assert" Version="$(XunitVersion)" />
+    <PackageReference Include="Humanizer.Core" Version="$(HumanizerVersion)"/>
   </ItemGroup>
   <Import Project="..\Utilities\Analyzer.Utilities.projitems" Label="Shared" />
 </Project>

--- a/src/Utilities/Extensions/StringExtensions.cs
+++ b/src/Utilities/Extensions/StringExtensions.cs
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Text;
 
 namespace Analyzer.Utilities.Extensions
 {
@@ -41,6 +42,12 @@ namespace Analyzer.Utilities.Extensions
             }
 
             return str.Substring(0, str.Length - suffix.Length);
+        }
+
+        public static bool IsASCII(this string value)
+        {
+            // ASCII encoding replaces non-ascii with question marks, so we use UTF8 to see if multi-byte sequences are there
+            return Encoding.UTF8.GetByteCount(value) == value.Length;
         }
     }
 }

--- a/src/Utilities/Extensions/StringExtensions.cs
+++ b/src/Utilities/Extensions/StringExtensions.cs
@@ -6,23 +6,6 @@ namespace Analyzer.Utilities.Extensions
 {
     internal static class StringExtensions
     {
-        /// <summary>
-        ///  Note: This doesnot handle all cases of plural for example "Men", "Vertices" etc
-        ///  Word ending with 'i' and 'ae' is to avoid irregular plurals 
-        /// </summary>
-        /// <param name="word"></param>
-        /// <returns></returns>
-        public static bool IsPlural(this string word)
-        {
-            if (!word.EndsWith("s", StringComparison.OrdinalIgnoreCase) &&
-                !word.EndsWith("i", StringComparison.OrdinalIgnoreCase) &&
-                !word.EndsWith("ae", StringComparison.OrdinalIgnoreCase))
-            {
-                return false;
-            }
-            return true;
-        }
-
         public static bool HasSuffix(this string str, string suffix)
         {
             if (str == null)

